docs/design/schemas.md

```diff
@@ -0,0 +1,180 @@
+
+### Both/either
+
+A `Resource` struct could have a `Schema` field, and a `GetSchema` field with func type added later if needed. The framework could use whichever is set, either erroring if both are set or using a documented fallback mechanism to prefer one over the other.
+
+## Types
+
+Whatever the relationship of schema to resource, it is structured data that must be represented by a Go struct or map type. This section details two options for this: the existing `Schema` type in `tfprotov6`, and a proposed `framework.Schema` type.
+
+### `tfprotov6.Schema`
+
+The `tfprotov6` (and `tfprotov5`) package in terraform-plugin-go provides a `Schema` type and related types: https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov6#Schema
+
+Resource schemas defined using these types can be quite verbose. For example, consider the following implementation of a resource with one attribute and one nested block, adapted from the `sql_migrate` resource from [terraform-provider-sql](https://github.com/paultyng/terraform-provider-sql/blob/main/internal/provider/data_query.go):
+
+```go
+var schema = &tfprotov5.Schema{
+	Block: &tfprotov5.SchemaBlock{
+		BlockTypes: []*tfprotov5.SchemaNestedBlock{
+			{
+				TypeName: "migration",
+				Nesting:  tfprotov5.SchemaNestedBlockNestingModeList,
+				Block: &tfprotov5.SchemaBlock{
+					Attributes: []*tfprotov5.SchemaAttribute{
+						{
+							Name:            "id",
+							Required:        true,
+							Description:     "Identifier can be any string to help identifying the migration in the source.",
+							DescriptionKind: tfprotov5.StringKindMarkdown,
+							Type:            tftypes.String,
+						},
+						{
+							Name:            "up",
+							Required:        true,
+							Description:     "The query to run when applying this migration.",
+							DescriptionKind: tfprotov5.StringKindMarkdown,
+							Type:            tftypes.String,
+						},
+						{
+							Name:            "down",
+							Required:        true,
+							Description:     "The query to run when undoing this migration.",
+							DescriptionKind: tfprotov5.StringKindMarkdown,
+							Type:            tftypes.String,
+						},
+					},
+				},
+			},
+		},
+		Attributes: []*tfprotov5.SchemaAttribute{
+			&tfprotov5.SchemaAttribute{
+				Name:     "complete_migrations",
+				Computed: true,
+				Description: "The completed migrations that have been run against your database. This list is used as " +
+					"storage to migrate down or as a trigger for downstream dependencies.",
+				DescriptionKind: tfprotov5.StringKindMarkdown,
+				Type: tftypes.List{
+					ElementType: tftypes.Object{
+						AttributeTypes: map[string]tftypes.Type{
+							"id":   tftypes.String,
+							"up":   tftypes.String,
+							"down": tftypes.String,
+						},
+					},
+				},
+			},
+		},
+	},
+}
+```
+
+### `framework.Schema`
```

Comment:
I think I have some nitpicking over what the `framework.Schema` type would look like, but I think the place to do that nitpicking and decide what complexity we could gloss over and what we need to surface could happen on the PR implementing that choice. I think my only ask here is that we just note that this example is illustrative and isn't necessarily the design that we'd ship, that we're just capturing the pattern of abstracting away some of the complexity. I only worry about it because it's such a detailed design. :)

Reply:
Indeed, this was the intention behind "A minimally verbose option could look something like" - I'll make this clearer.